firedrake/functionspacedata.py

```diff
@@ -81,67 +83,71 @@ def get_node_set(mesh, nodes_per_entity):
     return node_set
 
 
-def get_cell_node_list(mesh, entity_dofs, global_numbering):
+def get_cell_node_list(mesh, entity_dofs, global_numbering, offsets):
     """Get the cell->node list for specified dof layout.
 
     :arg mesh: The mesh to use.
     :arg entity_dofs: The FInAT entity_dofs dict.
     :arg global_numbering: The PETSc Section describing node layout
         (see :func:`get_global_numbering`).
+    :arg offsets: layer offsets for each entity (maybe ignored).
     :returns: A numpy array mapping mesh cells to function space
         nodes.
     """
-    return mesh.make_cell_node_list(global_numbering, entity_dofs)
+    return mesh.make_cell_node_list(global_numbering, entity_dofs, offsets)
 
 
-def get_facet_node_list(mesh, kind, cell_node_list):
+def get_facet_node_list(mesh, kind, cell_node_list, offsets):
     """Get the facet->node list for specified dof layout.
 
     :arg mesh: The mesh to use.
     :arg kind: The facet kind (one of ``"interior_facets"`` or
         ``"exterior_facets"``).
     :arg cell_node_list: The map from mesh cells to function space
         nodes, see :func:`get_cell_node_list`.
+    :arg offsets: layer offsets for each entity (maybe ignored).
     :returns: A numpy array mapping mesh facets to function space
         nodes.
     """
     assert kind in ["interior_facets", "exterior_facets"]
     if mesh._plex.getStratumSize(kind, 1) > 0:
-        facet = getattr(mesh, kind)
-        return dm_mod.get_facet_nodes(facet.facet_cell, cell_node_list)
+        return mesh.make_facet_node_list(cell_node_list, kind, offsets)
     else:
         return numpy.array([], dtype=IntType)
 
 
 @cached
-def get_entity_node_lists(mesh, key, entity_dofs, global_numbering):
+def get_entity_node_lists(mesh, key, entity_dofs, global_numbering, offsets):
     """Get the map from mesh entity sets to function space nodes.
 
     :arg mesh: The mesh to use.
     :arg key: Canonicalised entity_dofs (see :func:`entity_dofs_key`).
     :arg entity_dofs: FInAT entity dofs.
     :arg global_numbering: The PETSc Section describing node layout
         (see :func:`get_global_numbering`).
+    :arg offsets: layer offsets for each entity (maybe ignored).
     :returns: A dict mapping mesh entity sets to numpy arrays of
         function space nodes.
     """
     # set->node lists are specific to the sorted entity_dofs.
-    cell_node_list = get_cell_node_list(mesh, entity_dofs, global_numbering)
-    interior_facet_node_list = get_facet_node_list(mesh, "interior_facets", cell_node_list)
-    exterior_facet_node_list = get_facet_node_list(mesh, "exterior_facets", cell_node_list)
+    cell_node_list = get_cell_node_list(mesh, entity_dofs, global_numbering, offsets)
+    interior_facet_node_list = get_facet_node_list(mesh, "interior_facets", cell_node_list, offsets)
+    exterior_facet_node_list = get_facet_node_list(mesh, "exterior_facets", cell_node_list, offsets)
```

Comment:
This could be done like so, but it's quite magic.

```patch
diff --git a/firedrake/functionspacedata.py b/firedrake/functionspacedata.py
index 83408aa..bc336aa 100644
--- a/firedrake/functionspacedata.py
+++ b/firedrake/functionspacedata.py
@@ -17,7 +17,7 @@ vs VectorElement) can share the PyOP2 Set and Map data.
 import numpy
 import finat
 from decorator import decorator
-from functools import reduce
+from functools import reduce, partial
 
 from coffee import base as ast
 
@@ -144,11 +144,18 @@ def get_entity_node_lists(mesh, key, entity_dofs, global_numbering, offsets):
     """
     # set->node lists are specific to the sorted entity_dofs.
     cell_node_list = get_cell_node_list(mesh, entity_dofs, global_numbering, offsets)
-    interior_facet_node_list = get_facet_node_list(mesh, "interior_facets", cell_node_list, offsets)
-    exterior_facet_node_list = get_facet_node_list(mesh, "exterior_facets", cell_node_list, offsets)
-    return {mesh.cell_set: cell_node_list,
-            mesh.interior_facets.set: interior_facet_node_list,
-            mesh.exterior_facets.set: exterior_facet_node_list}
+    interior_facet_node_list = partial(get_facet_node_list, mesh, "interior_facets", cell_node_list, offsets)
+    exterior_facet_node_list = partial(get_facet_node_list, mesh, "exterior_facets", cell_node_list, offsets)
+
+    class magic(dict):
+        def __missing__(self, key):
+            return self.setdefault(mesh.interior_facets.set,
+                                   {mesh.interior_facets.set: interior_facet_node_list,
+                                    mesh.exterior_facets.set: exterior_facet_node_list}[key]())
+
+    retval = magic()
+    retval[mesh.cell_set] = cell_node_list
+    return retval
 
 
 @cached
```